File: src/common/error_handling.rs
// Syn_OS Standardized Error Handling Framework
// Provides unified error handling, logging, and recovery patterns for Rust components

use serde::{Deserialize, Serialize};
use std::collections::HashMap;
use std::fmt;
use std::error::Error as StdError;
use std::time::SystemTime;
use log::{error, warn, info, debug};

/// Standardized error severity levels
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
pub enum ErrorSeverity {
    Critical,  // System failure, requires immediate action
    High,      // Service degradation, user impact
    Medium,    // Functionality impaired, workaround available
    Low,       // Minor issues, no user impact
    Info,      // Informational, no action required
}

/// Standardized error categories
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize, Hash)]
pub enum ErrorCategory {
    Authentication,
    Authorization,
    Validation,
    Network,
    Database,
    Filesystem,
    Configuration,
    Consciousness,
    Integration,
    Security,
    Performance,
    System,
}

/// Structured error context for additional information
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ErrorContext {
    pub data: HashMap<String, String>,
}

impl ErrorContext {
    pub fn new() -> Self {
        Self {
            data: HashMap::new(),
        }
    }
    
    pub fn with(mut self, key: &str, value: &str) -> Self {
        self.data.insert(key.to_string(), value.to_string());
        self
    }
    
    pub fn add(&mut self, key: &str, value: &str) {
        self.data.insert(key.to_string(), value.to_string());
    }
}

impl Default for ErrorContext {
    fn default() -> Self {
        Self::new()
    }
}

/// Main error structure for Syn_OS
#[derive(Debug, Serialize, Deserialize)]
pub struct SynOSError {
    pub message: String,
    pub category: ErrorCategory,
    pub severity: ErrorSeverity,
    pub error_code: String,
    pub context: ErrorContext,
    pub timestamp: u64,
    pub service: String,
}

impl SynOSError {
    pub fn new(
        message: &str,
        category: ErrorCategory,
        severity: ErrorSeverity,
        service: &str,
    ) -> Self {
        let error_code = format!("{:?}_{:?}", category, severity).to_uppercase();
        let timestamp = SystemTime::now()
            .duration_since(SystemTime::UNIX_EPOCH)
            .unwrap_or_default()
            .as_secs();

        Self {
            message: message.to_string(),
            category,
            severity,
            error_code,
            context: ErrorContext::new(),
            timestamp,
            service: service.to_string(),
        }
    }
    
    pub fn with_context(mut self, context: ErrorContext) -> Self {
        self.context = context;
        self
    }
    
    pub fn add_context(mut self, key: &str, value: &str) -> Self {
        self.context.add(key, value);
        self
    }
    
    /// Convert error to JSON for structured logging
    pub fn to_json(&self) -> String {
        serde_json::to_string_pretty(self).unwrap_or_else(|_| self.message.clone())
    }
    
    /// Check if error is critical and requires immediate attention
    pub fn is_critical(&self) -> bool {
        matches!(self.severity, ErrorSeverity::Critical)
    }
    
    /// Get appropriate log level for this error
    pub fn log_level(&self) -> log::Level {
        match self.severity {
            ErrorSeverity::Critical => log::Level::Error,
            ErrorSeverity::High => log::Level::Error,
            ErrorSeverity::Medium => log::Level::Warn,
            ErrorSeverity::Low => log::Level::Info,
            ErrorSeverity::Info => log::Level::Debug,
        }
    }
}

impl fmt::Display for SynOSError {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(f, "[{}] {}: {}", self.error_code, self.service, self.message)
    }
}

impl StdError for SynOSError {}

/// Result type alias for Syn_OS operations
pub type SynOSResult<T> = Result<T, SynOSError>;

/// Error handler for centralized error management
pub struct ErrorHandler {
    service_name: String,
    error_stats: HashMap<ErrorCategory, u64>,
}

impl ErrorHandler {
    pub fn new(service_name: &str) -> Self {
        Self {
            service_name: service_name.to_string(),
            error_stats: HashMap::new(),
        }
    }
    
    /// Handle and log error with appropriate severity
    pub fn handle_error(&mut self, error: &SynOSError) {
        // Update statistics
        *self.error_stats.entry(error.category).or_insert(0) += 1;
        
        // Log with appropriate level
        match error.severity {
            ErrorSeverity::Critical => error!("[CRITICAL] {}", error.to_json()),
            ErrorSeverity::High => error!("[HIGH] {}", error.to_json()),
            ErrorSeverity::Medium => warn!("[MEDIUM] {}", error.to_json()),
            ErrorSeverity::Low => info!("[LOW] {}", error.to_json()),
            ErrorSeverity::Info => debug!("[INFO] {}", error.to_json()),
        }
        
        // Send alerts for critical errors
        if error.is_critical() {
            self.send_critical_alert(error);
        }
    }
    
    /// Send critical error alerts
    fn send_critical_alert(&self, error: &SynOSError) {
        // In a real implementation, this would integrate with alerting systems
        // For now, log to a special critical alerts file
        use std::fs::OpenOptions;
        use std::io::Write;
        
        if let Ok(mut file) = OpenOptions::new()
            .create(true)
            .append(true)
            .open("/home/diablorain/Syn_OS/logs/errors/critical_alerts.log")
        {
            let alert = format!("CRITICAL ALERT: {}\n", error.to_json());
            let _ = file.write_all(alert.as_bytes());
        }
    }
    
    /// Get error statistics for monitoring
    pub fn get_error_statistics(&self) -> HashMap<ErrorCategory, u64> {
        self.error_stats.clone()
    }
    
    /// Reset error statistics
    pub fn reset_statistics(&mut self) {
        self.error_stats.clear();
    }
}

/// Macro for creating specific error types
macro_rules! define_error_type {
    ($name:ident, $category:expr) => {
        pub fn $name(message: &str, service: &str) -> SynOSError {
            SynOSError::new(message, $category, ErrorSeverity::Medium, service)
        }
    };
}

// Define specific error constructors
define_error_type!(authentication_error, ErrorCategory::Authentication);
define_error_type!(authorization_error, ErrorCategory::Authorization);
define_error_type!(validation_error, ErrorCategory::Validation);
define_error_type!(network_error, ErrorCategory::Network);
define_error_type!(database_error, ErrorCategory::Database);
define_error_type!(filesystem_error, ErrorCategory::Filesystem);
define_error_type!(configuration_error, ErrorCategory::Configuration);
define_error_type!(consciousness_error, ErrorCategory::Consciousness);
define_error_type!(integration_error, ErrorCategory::Integration);
define_error_type!(security_error, ErrorCategory::Security);
define_error_type!(performance_error, ErrorCategory::Performance);
define_error_type!(system_error, ErrorCategory::System);

/// Trait for converting standard errors to SynOSError
pub trait IntoSynOSError {
    fn into_syn_os_error(self, category: ErrorCategory, service: &str) -> SynOSError;
}

impl<E: StdError> IntoSynOSError for E {
    fn into_syn_os_error(self, category: ErrorCategory, service: &str) -> SynOSError {
        SynOSError::new(&self.to_string(), category, ErrorSeverity::Medium, service)
            .add_context("original_error", &self.to_string())
    }
}

/// Extension trait for Result to add context and convert errors
pub trait ResultExt<T> {
    /// Add context to error if Result is Err
    fn with_context(self, key: &str, value: &str) -> SynOSResult<T>;
    
    /// Convert standard Result to SynOSResult with specified category
    fn to_syn_os_result(self, category: ErrorCategory, service: &str) -> SynOSResult<T>;
    
    /// Log error and continue with default value
    fn log_and_default(self, default: T, handler: &mut ErrorHandler) -> T;
}

impl<T, E: StdError> ResultExt<T> for Result<T, E> {
    fn with_context(self, key: &str, value: &str) -> SynOSResult<T> {
        self.map_err(|e| {
            e.into_syn_os_error(ErrorCategory::System, "unknown")
                .add_context(key, value)
        })
    }
    
    fn to_syn_os_result(self, category: ErrorCategory, service: &str) -> SynOSResult<T> {
        self.map_err(|e| e.into_syn_os_error(category, service))
    }
    
    fn log_and_default(self, default: T, handler: &mut ErrorHandler) -> T {
        match self {
            Ok(value) => value,
            Err(e) => {
                let syn_error = e.into_syn_os_error(ErrorCategory::System, "unknown");
                handler.handle_error(&syn_error);
                default
            }
        }
    }
}

impl<T> ResultExt<T> for SynOSResult<T> {
    fn with_context(self, key: &str, value: &str) -> SynOSResult<T> {
        self.map_err(|e| e.add_context(key, value))
    }
    
    fn to_syn_os_result(self, _category: ErrorCategory, _service: &str) -> SynOSResult<T> {
        self
    }
    
    fn log_and_default(self, default: T, handler: &mut ErrorHandler) -> T {
        match self {
            Ok(value) => value,
            Err(e) => {
                handler.handle_error(&e);
                default
            }
        }
    }
}

/// Safe execution function that catches panics and converts to errors
pub fn safe_execute<F, T>(f: F, service: &str) -> SynOSResult<T>
where
    F: FnOnce() -> T,
{
    match std::panic::catch_unwind(std::panic::AssertUnwindSafe(f)) {
        Ok(result) => Ok(result),
        Err(_) => Err(system_error("Function panicked during execution", service)
            .add_context("panic", "true")),
    }
}

/// Convenience macro for error handling in functions
#[macro_export]
macro_rules! handle_error {
    ($result:expr, $handler:expr) => {
        match $result {
            Ok(val) => val,
            Err(e) => {
                $handler.handle_error(&e);
                return Err(e);
            }
        }
    };
}

/// Convenience macro for logging errors without returning
#[macro_export]
macro_rules! log_error {
    ($result:expr, $handler:expr) => {
        if let Err(e) = $result {
            $handler.handle_error(&e);
        }
    };
}

#[cfg(test)]
mod tests {
    use super::*;
    
    #[test]
    fn test_error_creation() {
        let error = validation_error("Invalid input", "test_service")
            .add_context("input", "invalid_data");
        
        assert_eq!(error.category, ErrorCategory::Validation);
        assert_eq!(error.severity, ErrorSeverity::Medium);
        assert!(error.context.data.contains_key("input"));
    }
    
    #[test]
    fn test_error_handler() {
        let mut handler = ErrorHandler::new("test");
        let error = system_error("Test error", "test");
        
        handler.handle_error(&error);
        let stats = handler.get_error_statistics();
        
        assert_eq!(stats[&ErrorCategory::System], 1);
    }
    
    #[test]
    fn test_result_extensions() {
        let result: Result<i32, std::io::Error> = Err(std::io::Error::new(
            std::io::ErrorKind::NotFound,
            "File not found"
        ));
        
        let syn_result = result.to_syn_os_result(ErrorCategory::Filesystem, "test");
        assert!(syn_result.is_err());
        
        if let Err(e) = syn_result {
            assert_eq!(e.category, ErrorCategory::Filesystem);
        }
    }
}

File: src/common/error_handling.go
// Package errors provides standardized error handling for Syn_OS Go components
package errors

import (
	"encoding/json"
	"fmt"
	"log"
	"os"
	"path/filepath"
	"runtime"
	"time"
)

// ErrorSeverity defines the severity levels for errors
type ErrorSeverity string

const (
	SeverityCritical ErrorSeverity = "CRITICAL" // System failure, requires immediate action
	SeverityHigh     ErrorSeverity = "HIGH"     // Service degradation, user impact
	SeverityMedium   ErrorSeverity = "MEDIUM"   // Functionality impaired, workaround available
	SeverityLow      ErrorSeverity = "LOW"      // Minor issues, no user impact
	SeverityInfo     ErrorSeverity = "INFO"     // Informational, no action required
)

// ErrorCategory defines the categories of errors
type ErrorCategory string

const (
	CategoryAuthentication ErrorCategory = "AUTHENTICATION"
	CategoryAuthorization  ErrorCategory = "AUTHORIZATION"
	CategoryValidation     ErrorCategory = "VALIDATION"
	CategoryNetwork        ErrorCategory = "NETWORK"
	CategoryDatabase       ErrorCategory = "DATABASE"
	CategoryFilesystem     ErrorCategory = "FILESYSTEM"
	CategoryConfiguration  ErrorCategory = "CONFIGURATION"
	CategoryConsciousness  ErrorCategory = "CONSCIOUSNESS"
	CategoryIntegration    ErrorCategory = "INTEGRATION"
	CategorySecurity       ErrorCategory = "SECURITY"
	CategoryPerformance    ErrorCategory = "PERFORMANCE"
	CategorySystem         ErrorCategory = "SYSTEM"
)

// ErrorContext provides additional context for errors
type ErrorContext map[string]interface{}

// SynOSError represents a standardized error in Syn_OS
type SynOSError struct {
	Message       string        `json:"message"`
	Category      ErrorCategory `json:"category"`
	Severity      ErrorSeverity `json:"severity"`
	ErrorCode     string        `json:"error_code"`
	Context       ErrorContext  `json:"context"`
	Timestamp     time.Time     `json:"timestamp"`
	Service       string        `json:"service"`
	StackTrace    string        `json:"stack_trace,omitempty"`
	OriginalError error         `json:"original_error,omitempty"`
}

// Error implements the error interface
func (e *SynOSError) Error() string {
	return fmt.Sprintf("[%s] %s: %s", e.ErrorCode, e.Service, e.Message)
}

// NewError creates a new SynOSError
func NewError(message string, category ErrorCategory, severity ErrorSeverity, service string) *SynOSError {
	errorCode := fmt.Sprintf("%s_%s", category, severity)

	// Get stack trace
	buf := make([]byte, 1024)
	runtime.Stack(buf, false)

	return &SynOSError{
		Message:    message,
		Category:   category,
		Severity:   severity,
		ErrorCode:  errorCode,
		Context:    make(ErrorContext),
		Timestamp:  time.Now().UTC(),
		Service:    service,
		StackTrace: string(buf),
	}
}

// WithContext adds context to the error
func (e *SynOSError) WithContext(key string, value interface{}) *SynOSError {
	e.Context[key] = value
	return e
}

// WithOriginalError adds the original error that caused this error
func (e *SynOSError) WithOriginalError(err error) *SynOSError {
	e.OriginalError = err
	if e.Context == nil {
		e.Context = make(ErrorContext)
	}
	e.Context["original_error"] = err.Error()
	return e
}

// IsCritical returns true if the error is critical
func (e *SynOSError) IsCritical() bool {
	return e.Severity == SeverityCritical
}

// ToJSON returns the error as a JSON string
func (e *SynOSError) ToJSON() string {
	jsonBytes, err := json.MarshalIndent(e, "", "  ")
	if err != nil {
		return fmt.Sprintf(`{"error": "Failed to marshal error to JSON: %s", "original_message": "%s"}`, err.Error(), e.Message)
	}
	return string(jsonBytes)
}

// ErrorHandler manages error handling and logging for a service
type ErrorHandler struct {
	ServiceName  string
	LogFile      *os.File
	ErrorStats   map[ErrorCategory]int64
	AlertChannel chan *SynOSError
}

// NewErrorHandler creates a new error handler
func NewErrorHandler(serviceName string) (*ErrorHandler, error) {
	// Create logs directory if it doesn't exist
	logDir := "/home/diablorain/Syn_OS/logs/errors"
	if err := os.MkdirAll(logDir, 0755); err != nil {
		return nil, fmt.Errorf("failed to create log directory: %w", err)
	}

	// Open log file
	logFile, err := os.OpenFile(
		filepath.Join(logDir, fmt.Sprintf("%s_errors.log", serviceName)),
		os.O_CREATE|os.O_WRONLY|os.O_APPEND,
		0644,
	)
	if err != nil {
		return nil, fmt.Errorf("failed to open log file: %w", err)
	}

	handler := &ErrorHandler{
		ServiceName:  serviceName,
		LogFile:      logFile,
		ErrorStats:   make(map[ErrorCategory]int64),
		AlertChannel: make(chan *SynOSError, 100), // Buffered channel for alerts
	}

	// Start alert processor
	go handler.processAlerts()

	return handler, nil
}

// HandleError processes and logs an error
func (h *ErrorHandler) HandleError(err *SynOSError) {
	// Update statistics
	h.ErrorStats[err.Category]++

	// Log error
	logEntry := fmt.Sprintf("%s - %s\n", time.Now().UTC().Format(time.RFC3339), err.ToJSON())
	if _, writeErr := h.LogFile.WriteString(logEntry); writeErr != nil {
		log.Printf("Failed to write error to log file: %v", writeErr)
	}

	// Log to standard logger based on severity
	switch err.Severity {
	case SeverityCritical, SeverityHigh:
		log.Printf("ERROR: %s", err.Error())
	case SeverityMedium:
		log.Printf("WARN: %s", err.Error())
	case SeverityLow, SeverityInfo:
		log.Printf("INFO: %s", err.Error())
	}

	// Send alert for critical errors
	if err.IsCritical() {
		select {
		case h.AlertChannel <- err:
		default:
			log.Printf("Alert channel full, dropping critical alert: %s", err.Error())
		}
	}
}

// processAlerts handles critical error alerts
func (h *ErrorHandler) processAlerts() {
	alertFile, err := os.OpenFile(
		"/home/diablorain/Syn_OS/logs/errors/critical_alerts.log",
		os.O_CREATE|os.O_WRONLY|os.O_APPEND,
		0644,
	)
	if err != nil {
		log.Printf("Failed to open critical alerts file: %v", err)
		return
	}
	defer alertFile.Close()

	for alert := range h.AlertChannel {
		alertEntry := fmt.Sprintf("%s - CRITICAL ALERT: %s\n",
			time.Now().UTC().Format(time.RFC3339), alert.ToJSON())

		if _, err := alertFile.WriteString(alertEntry); err != nil {
			log.Printf("Failed to write critical alert: %v", err)
		}

		// Here you could integrate with external alerting systems
		// like PagerDuty, Slack, email, etc.
		log.Printf("🚨 CRITICAL ALERT: %s", alert.Error())
	}
}

// GetErrorStatistics returns error statistics
func (h *ErrorHandler) GetErrorStatistics() map[ErrorCategory]int64 {
	stats := make(map[ErrorCategory]int64)
	for category, count := range h.ErrorStats {
		stats[category] = count
	}
	return stats
}

// ResetStatistics resets error statistics
func (h *ErrorHandler) ResetStatistics() {
	h.ErrorStats = make(map[ErrorCategory]int64)
}

// Close closes the error handler and its resources
func (h *ErrorHandler) Close() error {
	close(h.AlertChannel)
	return h.LogFile.Close()
}

// Convenience functions for creating specific error types

// AuthenticationError creates an authentication error
func AuthenticationError(message, service string) *SynOSError {
	return NewError(message, CategoryAuthentication, SeverityHigh, service)
}

// AuthorizationError creates an authorization error
func AuthorizationError(message, service string) *SynOSError {
	return NewError(message, CategoryAuthorization, SeverityHigh, service)
}

// ValidationError creates a validation error
func ValidationError(message, service string) *SynOSError {
	return NewError(message, CategoryValidation, SeverityMedium, service)
}

// NetworkError creates a network error
func NetworkError(message, service string) *SynOSError {
	return NewError(message, CategoryNetwork, SeverityMedium, service)
}

// DatabaseError creates a database error
func DatabaseError(message, service string) *SynOSError {
	return NewError(message, CategoryDatabase, SeverityHigh, service)
}

// FilesystemError creates a filesystem error
func FilesystemError(message, service string) *SynOSError {
	return NewError(message, CategoryFilesystem, SeverityMedium, service)
}

// ConfigurationError creates a configuration error
func ConfigurationError(message, service string) *SynOSError {
	return NewError(message, CategoryConfiguration, SeverityHigh, service)
}

// ConsciousnessError creates a consciousness error
func ConsciousnessError(message, service string) *SynOSError {
	return NewError(message, CategoryConsciousness, SeverityCritical, service)
}

// IntegrationError creates an integration error
func IntegrationError(message, service string) *SynOSError {
	return NewError(message, CategoryIntegration, SeverityMedium, service)
}

// SecurityError creates a security error
func SecurityError(message, service string) *SynOSError {
	return NewError(message, CategorySecurity, SeverityCritical, service)
}

// PerformanceError creates a performance error
func PerformanceError(message, service string) *SynOSError {
	return NewError(message, CategoryPerformance, SeverityLow, service)
}

// SystemError creates a system error
func SystemError(message, service string) *SynOSError {
	return NewError(message, CategorySystem, SeverityMedium, service)
}

// Wrap wraps a standard error into a SynOSError
func Wrap(err error, category ErrorCategory, service string) *SynOSError {
	if err == nil {
		return nil
	}

	synError := NewError(err.Error(), category, SeverityMedium, service)
	return synError.WithOriginalError(err)
}

// WrapWithMessage wraps a standard error with a custom message
func WrapWithMessage(err error, message string, category ErrorCategory, service string) *SynOSError {
	if err == nil {
		return nil
	}

	synError := NewError(message, category, SeverityMedium, service)
	return synError.WithOriginalError(err)
}

// Safe execution functions

// SafeExecute executes a function and converts any panic to a SynOSError
func SafeExecute(fn func() error, service string) (err *SynOSError) {
	defer func() {
		if r := recover(); r != nil {
			err = SystemError(fmt.Sprintf("Function panicked: %v", r), service).
				WithContext("panic", true)
		}
	}()

	if fnErr := fn(); fnErr != nil {
		if synErr, ok := fnErr.(*SynOSError); ok {
			return synErr
		}
		return Wrap(fnErr, CategorySystem, service)
	}

	return nil
}

// LogAndContinue logs an error and continues execution
func LogAndContinue(err error, handler *ErrorHandler, service string) {
	if err == nil {
		return
	}

	var synErr *SynOSError
	if se, ok := err.(*SynOSError); ok {
		synErr = se
	} else {
		synErr = Wrap(err, CategorySystem, service)
	}

	handler.HandleError(synErr)
}

// MustNotError panics if error is not nil (for initialization code)
func MustNotError(err error, message string) {
	if err != nil {
		panic(fmt.Sprintf("%s: %v", message, err))
	}
}

// Example usage and testing
func Example() {
	// Create error handler
	handler, err := NewErrorHandler("example_service")
	if err != nil {
		log.Fatal("Failed to create error handler:", err)
	}
	defer handler.Close()

	// Create and handle different types of errors
	authErr := AuthenticationError("Invalid credentials", "auth_service").
		WithContext("user_id", "12345").
		WithContext("ip_address", "192.168.1.100")

	handler.HandleError(authErr)

	// Wrap a standard error
	_, openErr := os.Open("/nonexistent/file")
	if openErr != nil {
		synErr := Wrap(openErr, CategoryFilesystem, "file_service").
			WithContext("operation", "open_file").
			WithContext("file_path", "/nonexistent/file")

		handler.HandleError(synErr)
	}

	// Safe execution example
	if execErr := SafeExecute(func() error {
		// This would panic
		panic("Something went wrong")
	}, "test_service"); execErr != nil {
		handler.HandleError(execErr)
	}

	// Print statistics
	stats := handler.GetErrorStatistics()
	for category, count := range stats {
		log.Printf("Category %s: %d errors", category, count)
	}
}
